### PR TITLE
UWP rework

### PR DESCRIPTION
**Pull request by Rover656**
Here is the ~~first draft of the~~ UWP rework discussed in discord and relating to issue #1182 

What I have done:
- Replaced the UWP messaging system that I added last time with a function callback system.
- The developer is made to implement all UWP support now (for window and application lifecycle only).
- The new system supports more of raylib's core.

What I have tested:
- Keyboard input
- Mouse input
- Window lifecycle

What I am still to test:
- [x] Gamepad input
- [x] Keyboard character recording
- [x] Whether or not UWP will support GIF recording.

What I am still to add/do:
- [x] Move UWP specifics to a separate header.
- [x] Add keyboard character recording.
- [x] I still need to work out if locking the mouse cursor position is truly possible or if the developer will have to do it themselves. This would be easy as they can set their own functions for this anyway.
- [x] Move BaseApp into the main App. This is now less of a boilerplate and more of an example
- [x] Add full set of keyboard keys to example
- [x] Try to fix the gamepad hack
- [x] Implement UWP lifecycle (Resume and Suspend) properly.
- [x] Ensure code matches raylib standards.

### CODE REVIEW

**Comment by raysan5** [comment]
@Rover656 this is a big batch of changes... looking at the code it probably makes sense to move all UWP specific functionality to a self-contained `uwp_events.h` external library, used only on `core.c`.

Also, where is done the events polling? is it done by the App?

**Comment by Rover656** [comment]
Hi ray, I agree with the moving of the UWP functionality to another header. Also, the event polling is only done by the App now, as most if not all of the functionality requires c++ access.

**Comment by Rover656** [comment]
I believe that this is now ready for reviewing.

**Comment by Rover656** [comment]
I'm going to look into UWP touch support, however I'm unsure how compatible it will be with raylib's setup as I've never used the touch APIs.

File: src/core.c [hunk 1]
```diff
@@ -2673,19 +2687,19 @@ int GetMouseWheelMove(void)
 // Returns touch position X for touch point 0 (relative to screen size)
 int GetTouchX(void)
 {
-#if defined(PLATFORM_ANDROID) || defined(PLATFORM_WEB)
+#if defined(PLATFORM_ANDROID) || defined(PLATFORM_WEB) || defined(PLATFORM_UWP)
     return (int)CORE.Input.Touch.position[0].x;
-#else   // PLATFORM_DESKTOP, PLATFORM_RPI
+#else   // PLATFORM_DESKTOP, PLATFORM_RPI, PLATFORM_UWP
```
Comment (raysan5) [inline src/core.c:2692]: Just a typo I think, if PLATFORM_UWP is considered on previous #if, it shouldn't be here in the comment

**Comment by raysan5** [comment]
@Rover656 I looks very nice! Congratulations! I like it! 😄

File: src/uwp_events.h [hunk 2]
```diff
@@ -0,0 +1,119 @@
+/**********************************************************************************************
+*
+*   raylib.uwp_events - Functions for bootstrapping UWP functionality within raylib's core.
+*
+*
+*   LICENSE: zlib/libpng
+*
+*   Copyright (c) 2014-2020 Ramon Santamaria (@raysan5)
```
Comment (raysan5) [inline src/uwp_events.h:8]: You must be the one in this copyright notice, you created it!

File: src/core.c (hunk 1 above)
Comment (Rover656) [reply src/core.c:2692]: Oh, sorry. I thought that was talking about what the if contained. I shall remove these

File: src/uwp_events.h (hunk 2 above)
Comment (Rover656) [reply src/uwp_events.h:8]: Alrighty, I shall add that too

**Comment by raysan5** [comment]
@Rover656 Great improvement! Thank you very much!

**Pull request merged**